lib/widgets/login.dart

```diff
@@ -110,22 +120,34 @@ class _EmailPasswordLoginPageState extends State<EmailPasswordLoginPage> {
   void _submit() async {
     final context = _emailKey.currentContext!;
     final realmUrl = widget.serverSettings.realmUri;
-    final String? email = _emailKey.currentState!.value;
-    final String? password = _passwordKey.currentState!.value;
-    if (email == null || password == null) {
-      // TODO can these FormField values actually be null? when?
+    final emailFieldState = _emailKey.currentState!;
+    final passwordFieldState = _passwordKey.currentState!;
+    final emailValid = emailFieldState.validate(); // Side effect: on-field error text
+    final passwordValid = passwordFieldState.validate(); // Side effect: on-field error text
+    if (!emailValid || !passwordValid) {
       return;
     }
-    // TODO(#35): validate email is in the shape of an email
+    final String email = emailFieldState.value!;
+    final String password = passwordFieldState.value!;
 
     final FetchApiKeyResult result;
+    setState(() {
+      _inProgress = true;
+    });
     try {
       result = await fetchApiKey(
         realmUrl: realmUrl, username: email, password: password);
-    } on Exception catch (e) { // TODO(#37): distinguish API exceptions
-      // TODO(#35): give feedback to user on failed login
-      debugPrint(e.toString());
+    } on Exception { // TODO(#37): distinguish API exceptions
+      if (!context.mounted) return;
+      // TODO(#35) give more helpful feedback. Needs #37. The RN app is
+      //   unhelpful here; we should at least recognize invalid auth errors, and
+      //   errors for deactivated user or realm (see zulip-mobile#4571).
+      showErrorDialog(context: context, title: 'Login failed');
       return;
+    } finally {
+      setState(() {
+        _inProgress = false;
```

Comment:
Similar story to https://github.com/zulip/zulip-flutter/pull/87#discussion_r1185670796 ­— put this at the end of the function, and await the navigation.